core/src/avm2/amf.rs

```diff
@@ -55,22 +55,27 @@ pub fn serialize_value<'gc>(
                 // Don't serialize properties from the vtable (we don't want a 'length' field)
                 recursive_serialize(activation, o, &mut values, None, amf_version, object_table)
                     .unwrap();
+                let len = o.as_array_storage().unwrap().length() as u32;
 
-                let mut dense = vec![];
-                let mut sparse = vec![];
-                // ActionScript `Array`s can have non-number properties, and these properties
-                // are confirmed and tested to also be serialized, so do not limit the values
-                // iterated over by the length of the internal array data.
-                for (i, elem) in values.into_iter().enumerate() {
-                    if elem.name == i.to_string() {
-                        dense.push(elem.value.clone());
-                    } else {
-                        sparse.push(elem);
+                if amf_version == AMFVersion::AMF3 {
+                    let mut dense = vec![];
+                    let mut sparse = vec![];
+                    // ActionScript `Array`s can have non-number properties, and these properties
+                    // are confirmed and tested to also be serialized, so do not limit the values
+                    // iterated over by the length of the internal array data.
+                    for (i, elem) in values.into_iter().enumerate() {
+                        if elem.name == i.to_string() {
+                            dense.push(elem.value.clone());
+                        } else {
+                            sparse.push(elem);
+                        }
                     }
-                }
 
-                let len = o.as_array_storage().unwrap().length() as u32;
-                Some(AmfValue::ECMAArray(dense, sparse, len))
+                    Some(AmfValue::ECMAArray(dense, sparse, len))
+                } else {
+                    // TODO: is this right?
```

Comment:
Mhm did https://github.com/ruffle-rs/rust-flash-lso/issues/54 not make this work in AMF0?

Reply:
That's for writing down to bytes, but we don't do that here.

(Writing down to bytes has a whole host of other issues, for AMF3-in-AMF0. Rather avoid the cost of that if we don't need to, as well)